p2p/server.py

```diff
@@ -0,0 +1,157 @@
+            port=self.server_address.tcp_port,
+        )
+
+    async def run(self) -> None:
+        await self.start()
+        self.logger.info("Running server...")
+        await self.cancel_token.wait()
+        await self.stop()
+
+    async def stop(self) -> None:
+        self.logger.info("Closing server...")
+        self.cancel_token.trigger()
+        self._server.close()
+        await self._server.wait_closed()
+        await self.peer_pool.stop()
+
+    async def receive_handshake(
+            self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
+        # Use reader to read the auth_init msg until EOF
+        msg = await reader.read(ENCRYPTED_AUTH_MSG_LEN)
+
+        # Use HandshakeResponder.decode_authentication(auth_init_message) on auth init msg
+        try:
+            ephem_pubkey, initiator_nonce, initiator_pubkey = decode_authentication(
+                msg, self.privkey)
+        # Try to decode as EIP8
+        except DecryptionError:
+            msg_size = big_endian_to_int(msg[:2])
+            remaining_bytes = msg_size - ENCRYPTED_AUTH_MSG_LEN + 2
+            msg += await reader.read(remaining_bytes)
+            ephem_pubkey, initiator_nonce, initiator_pubkey = decode_authentication(
+                msg, self.privkey)
+
+        # Get remote's address: IPv4 or IPv6
+        ip, port, *_ = writer.get_extra_info("peername")
+        remote_address = Address(ip, port)
+
+        # Create `HandshakeResponder(remote: kademlia.Node, privkey: datatypes.PrivateKey)` instance
+        initiator_remote = Node(initiator_pubkey, remote_address)
+        responder = HandshakeResponder(initiator_remote, self.privkey)
+
+        # Call `HandshakeResponder.create_auth_ack_message(nonce: bytes)` to create the reply
+        responder_nonce = secrets.token_bytes(HASH_LEN)
+        auth_ack_msg = responder.create_auth_ack_message(nonce=responder_nonce)
+        auth_ack_ciphertext = responder.encrypt_auth_ack_message(auth_ack_msg)
+
+        # Use the `writer` to send the reply to the remote
+        writer.write(auth_ack_ciphertext)
+        await writer.drain()
+
+        # Call `HandshakeResponder.derive_shared_secrets()` and use return values to create `Peer`
+        aes_secret, mac_secret, egress_mac, ingress_mac = responder.derive_secrets(
+            initiator_nonce=initiator_nonce,
+            responder_nonce=responder_nonce,
+            remote_ephemeral_pubkey=ephem_pubkey,
+            auth_init_ciphertext=msg,
+            auth_ack_ciphertext=auth_ack_ciphertext
+        )
+
+        # Create and register peer in peer_pool
+        eth_peer = ETHPeer(
+            remote=initiator_remote, privkey=self.privkey, reader=reader,
+            writer=writer, aes_secret=aes_secret, mac_secret=mac_secret,
+            egress_mac=egress_mac, ingress_mac=ingress_mac, chaindb=self.chaindb,
+            network_id=self.network_id
+        )
+        self.peer_pool.add_peer(eth_peer)
```

Comment:
I'm writing a test to make two client connection, maybe we should add a line after here?
```python
await eth_peer.do_p2p_handshake()
```

Reply:
Yeah, and you must catch `HandshakeFailure` and skip adding the peer to the pool in that case

Reply:
Could we merge this PR now, and then I create another PR for the patch with tests?

Reply:
@hwwhww :shipit: